halo/attest/keeper/keeper.go

```diff
@@ -209,15 +209,30 @@ func (k *Keeper) Approve(ctx context.Context, valset *cmttypes.ValidatorSet) err
 		valsByPower[addr] = val.VotingPower
 	}
 
-	skip := make(map[uint64]bool)           // Skip processing chains as soon as a pending attestation is found.
-	edgesByChain := make(map[uint64]uint64) // Track new minimum edges for updated vote windows.
+	approvedByChain := make(map[uint64]uint64) // Cache the latest approved attestation by chain.
 	for iter.Next() {
 		att, err := iter.Value()
 		if err != nil {
 			return errors.Wrap(err, "value")
 		}
-		if skip[att.GetChainId()] {
-			continue
+
+		// Ensure we approve sequentially, not skipping any heights.
+		{
+			// Populate the cache if not already.
+			if _, ok := approvedByChain[att.GetChainId()]; ok {
```

Comment:
Is there a reason why our linters require us to use the proto methods instead of call fields directly? For example here can't we just use `att.ChainId` instead of having to call `att.GetChainId()`? The latter is particularly annoying locally when debugging there is one more func to step into.

Reply:
The linters don't require this. It is personal choice. Not sure we should optimise for debuggers.

Reply:
Well, when we have a function call that we want to step into during debugging and the arguments we pass come from proto structs we step into the proto getters first and this extremely annoying.
Also, during debugging sessions, if we use the struct fields we can directly see the value of the struct field with a mouse hover over, whereas, with a getter method this is not possible and we have to actually call it first in order to see the value.

I just tried changing a random line from:

```go
edgesByChain[att.GetChainId()] = uintSub(att.GetHeight(), k.voteWindow)
```

to

```go
edgesByChain[att.ChainId] = uintSub(att.Height, k.voteWindow)
```
and then ran `golangci-lint run` and it returned it back to the original code. So, indeed, it is enforced by the linter(s) : (